### PR TITLE
feat: streaming read selective batches in a single Lance data file

### PR DESCRIPTION
**Pull request by eddyxu**
Provide a Stream to read each batch on disk from a single file.

### CODE REVIEW

File: rust/lance/src/io/stream.rs [hunk 1]
```diff
@@ -12,14 +12,42 @@
...
     /// Returns the schema of the stream.
     fn schema(&self) -> SchemaRef;
 }
+
+pub struct BoxedRecordBatchStream<'a> {
+    inner: BoxStream<'a, Result<RecordBatch>>,
+    schema: SchemaRef,
+}
+
+impl<'a> BoxedRecordBatchStream<'a> {
```
Comment (wjones127) [inline rust/lance/src/io/stream.rs:35]: You should check out DataFusion's [RecordBatchStreamAdapter](https://docs.rs/datafusion/31.0.0/src/datafusion/physical_plan/stream.rs.html#249-297). It does the same thing, but is generic over the stream, so can also handle unboxed cases.

One thing it does that's not here is supplying `size_hint()`, so I'd at least add that.

File: rust/lance/src/io/reader.rs [hunk 2]
```diff
@@ -373,6 +375,42 @@ impl FileReader {
...
+        let projection = Arc::new(projection);
+        let batches = (0..self.num_batches() as i32)
+            .filter(predicate)
+            .collect::<Vec<_>>();
+        let this = self;
+        let proj = projection.clone();
+        let inner = stream::iter(batches)
+            .zip(stream::repeat_with(move || proj.clone()))
+            .map(move |(batch_id, proj)| async move {
+                let proj = proj.clone();
```
Comment (wjones127) [inline rust/lance/src/io/reader.rs:403]: Don't we already have an owned copy of `proj` from the line above?

File: rust/lance/src/io/reader.rs [hunk 3]
```diff
@@ -373,6 +375,42 @@ impl FileReader {
...
+    pub(crate) fn batches_stream<'a>(
+        &'a self,
+        projection: Schema,
+        predicate: impl FnMut(&i32) -> bool + Send + Sync + 'a,
+    ) -> impl RecordBatchStream + 'a {
+        // Make projection an Arc so we can clone it and pass between threads.
+        let projection = Arc::new(projection);
+        let batches = (0..self.num_batches() as i32)
+            .filter(predicate)
+            .collect::<Vec<_>>();
```
Comment (wjones127) [inline rust/lance/src/io/reader.rs:397]: There any reason to collect the batches? I think we can just pass the iterator straight to `stream::iter()`.

**Review by wjones127 — CHANGES_REQUESTED** [review]

File: rust/lance/src/io/stream.rs (hunk 1 above)
Comment (eddyxu) [reply rust/lance/src/io/stream.rs:35]: Ok, I can implement that, similar to our `RecordBatchStream` which returns our `Result<Batch, lance::Error>`

File: rust/lance/src/io/reader.rs (hunk 2 above)
Comment (eddyxu) [reply rust/lance/src/io/reader.rs:403]: 🆗

**Review by wjones127 — APPROVED** [review]
Excellent! 🚀

**Pull request merged**